torch/_inductor/fx_passes/split_cat.py

```diff
@@ -189,9 +192,385 @@ def merge_splits(
     counters["inductor"]["consecutive_split_merged"] += 1
 
 
+def match_node(node, target_names, op="call_function"):
+    return node.op == op and node.target in target_names
+
+
+class MergeSplitCat(PatternEntry):
+    def apply(self, match: Match, graph: torch.fx.Graph, node: torch.fx.Node):
+        split_nodes = [
+            getitem_node.args[0]
+            for getitem_node in get_arg_value(node, 0, "tensors")
+            if match_node(getitem_node, {operator.getitem})
+            and match_node(getitem_node.args[0], {torch.split})
+        ]
+        if len(set(split_nodes)) == 0:
+            return
+        split_node = split_nodes[0]
+        split_sections = split_node.args[1]
+        if not isinstance(split_sections, (list, tuple)):  # Unnormalized split
+            return
+        # Find the next users (i.e. users after the getitem)
+        next_users = self.find_next_users(split_node)
+        if not next_users:
+            return
+        # Gather inputs of the next users. When inputs come from `split_node`, they are instead represented by
+        # a tuple indicating the split ranges. See `get_user_input_list` for more details
+        user_inputs_list = self.get_user_input_list(split_node, next_users)
+        # Simplify the split_sections based on user_inputs_list. In simpler cases, len(simplified_split_ranges) == 1 and
+        # we can simply replace the split node. Otherwise, we simplify it.
+        simplified_split_ranges = self.get_simplified_split_ranges(
+            split_sections, next_users, user_inputs_list
+        )
+        if not simplified_split_ranges:
+            return
+        transform_params_list = self.get_transform_params(
+            split_node, next_users, user_inputs_list
+        )
+        if not transform_params_list:
+            return
+
+        # Start actual replacement
+        user_inputs_list_new = self.replace_split(
+            graph, split_node, user_inputs_list, simplified_split_ranges
+        )
+        self.replace_cat(
+            graph, split_node, next_users, user_inputs_list_new, transform_params_list
+        )
+        self.erase_old_nodes(graph, split_node, next_users)
+
+    def find_next_users(self, split_node):
+        next_users = set()
```

Comment:
Sets are not determinstic in Python.

Reply:
The order of `next_users` doesn't matter. `next_users` are essentially things like `cat`, `stack` nodes (not getitems), so we can replace them in any order

Reply:
Replacing them in random order will make the output code of `torch.compile()` non-determinstic.  We consider that a bug.